File: examples/ros2-pubsub/py_pubsub/py_pubsub/subscriber.py
# Copyright (c) 2026 WHEELTEC ROS Textbook Contributors
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.

"""Receive textbook messages from a fixed ROS 2 topic."""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class CourseSubscriber(Node):
    """Log each message received from the course topic."""

    def __init__(self) -> None:
        super().__init__("course_py_listener")
        self.subscription = self.create_subscription(
            String,
            "/course/chatter",
            self.receive_message,
            10,
        )

    def receive_message(self, message: String) -> None:
        """Log a received string message."""
        self.get_logger().info(f"Received: '{message.data}'")


def main(args=None) -> None:
    """Run the subscriber until interrupted."""
    rclpy.init(args=args)
    node = CourseSubscriber()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: examples/ros2-pubsub/cpp_pubsub/src/subscriber.cpp
// Copyright (c) 2026 WHEELTEC ROS Textbook Contributors
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file or at
// https://opensource.org/licenses/MIT.

#include <functional>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using std::placeholders::_1;

class CourseSubscriber : public rclcpp::Node
{
public:
  CourseSubscriber()
  : Node("course_cpp_listener")
  {
    subscription_ = create_subscription<std_msgs::msg::String>(
      "/course/chatter", 10,
      std::bind(&CourseSubscriber::receive_message, this, _1));
  }

private:
  void receive_message(const std_msgs::msg::String & message) const
  {
    RCLCPP_INFO(get_logger(), "Received: '%s'", message.data.c_str());
  }

  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CourseSubscriber>());
  rclcpp::shutdown();
  return 0;
}
